### DISCUSSION TITLE
Updating listener clients from the UI

### DISCUSSION THREAD

**Discussion by MatteoGheza** [josephdadams/TallyArbiter#270] [Ideas]
I think that we can use [https://esphome.github.io/esp-web-tools/]() for updating listener clients, since it's production-ready, it's used by other awesome FOSS projects and it's maintained by NabuCasa.
I tested this and it's working well.
Example manifest file:
```js
{
  "name": "Tally Arbiter M5Atom Listener Client",
  "builds": [
    {
      "chipFamily": "ESP32",
      "improv": false,
      "parts": [
        {
          "path": "tallyarbiter-m5atom.ino.bin",
          "offset": 0
        }
      ]
    }
  ]
}
```

@hrueger what do you think about using this lib?

**Comment by josephdadams** [comment]
Wow. This is cool.

**Comment by JTF4** [comment]
What would you think about making the flashing software a separate component? That way you don't have to connect to the server to flash clients. (For people who may be running an off site or RPi server)

**Reply by hrueger** [reply]
@MatteoGheza The lib looks great.
@JTF4 I'd prefer to add the functionality to TallyArbiter and not create a separate application.
Then, we only have one app to maintain, people who use the desktop version can update their clients without installing another app and those who only have a remote server running will need to download something anyway - why not TallyArbiter again?

Edit: I just took a closer look at the lib and realized that the update works with WebUSB directly in the browser. That means that you only need to connect the listener client to a computer where you can access the web UI. That works with an instance running on a pi, too. You don't need to connect the client to the pi physically.

**Reply by JTF4** [reply]
Well if you can do it from the browser then I'm all for that.

**Reply by hrueger** [reply]
I just saw electron/electron#14545 . But as a workaround until that issue is fixed, we can detect if we're in the electron window and then open a Chrome window.

**Reply by MatteoGheza** [reply]
I think we can use https://thegecko.github.io/webusb/docs/index.html in Electron

**Comment by wrldtvlr** [comment]
Excited to see the release of TA 3.0 imminent, but a bit confused. We're running TA Server on a Pi 4 with four M5 Atom Matrix clients. I've read that the clients have been updated to work with TA 3.0, but I don't see any code changes in the M5 Matrix repository. I've also seen some discussion about embedding the client update process in the new TA 3.0 server, but it's not clear how that would be initiated. Through the server web interface?

**Comment by josephdadams** [comment]
All of the individual repositories for listener clients are going away and they are now all a part of the primary repo. Github will build binary files for us that you can flash directly to your atom. It’s covered in the documentation once we merge all those PR’s.

**Reply by josephdadams** [reply]
Just FYI I've marked all of the various listener client repositories under my Github as public archive and added a note to the readme stating that those versions are only compatible with TA < 3.0 and that all future requests, bug reports, and PR's should be made against this main repo.

**Comment by wrldtvlr** [comment]
Hmmm. There have been some significant code changes to the Matrix client from v2 to v3. I'm trying to compile the v3 info file in the Arduino IDE, but not getting past the include for WiFiManager.h. I suspect I may not have grabbed the correct WiFiManager library (which wasn't in the V2 client), as I get an undefined reference to ESP8266WiFi.h, which it shouldn't be calling as the Matrix is an ESP32. Any pointers to the correct libraries to load?

**Reply by wrldtvlr** [reply]
Disregard. Must have had some outdated libraries. Updated those and it compiled clean.

**Comment by wrldtvlr** [comment]
Next problem - Matrix client works, but the colors are really strange. The background is kind of a lavender color.  I think what's happening is that the server "bus" colors are being used, but while the names of those are fine (red & green), the hex representation is nothing like what the matrix needs or is defined in the client code.

I definitely could be misunderstanding how 
![IMG_20211129_191847](https://user-images.githubusercontent.com/8150022/143975732-e5dc8eda-0b54-43ed-bded-c8a6fe4ddf29.jpg)
colors work in V3 though
![IMG_20211129_191815](https://user-images.githubusercontent.com/8150022/143975739-3a2512f5-6ee0-4097-b921-183271c63ebe.jpg)
.

**Reply by mg-1999** [reply]
You could change the server side hex values to more vibrant colors, that should do the trick

**Reply by hrueger** [reply]
Yes, the default bus colors have been changed from bright red and green (#ff0000 and #00ff00) to this slightly more pale version. When you have the web tally client on a huge monitor, those bright colors really hurt in the eyes if you look at them for more than a few seconds.

If you change them in the settings, the updated listener clients of v3.0.0 should use the new colors. I don't know why there are still colors defined in the source code of the clients.

**Reply by mg-1999** [reply]
The colors defined in the listener clients are there for internal icons and text.

For example the m5atom uses the colors defined there for the startup animation and the connected to WiFi part.

The tally colors only get sent when there is an update, so they aren't available on startup yet

**Reply by hrueger** [reply]
Makes sense 👍 Thanks

**Reply by wrldtvlr** [reply]
I'll play with it in a few hours. It seems the only way to charge the color is with the picker, and not direct entry of the hex color number.

**Reply by wrldtvlr** [reply]
Sigh. Disregard again. Looked a little closer in the color picker and found how to enter the hex values. Working fine now.

**Reply by wrldtvlr** [reply]
Little more testing - I think the Pgm and Pre colors are flipped. The web interface shows Red for Pgm, and Green for Preview, as expected. The M5 colors are reversed.

**Reply by wrldtvlr** [reply]
Just in case anyone misses the other thread. Red & green were reversed in the draw pixel call. Corrected in latest build.
